### PR TITLE
feat(smart-contracts): tests with uniswap v3 oracle 

### PR DESCRIPTION
**Pull request by clemsos**
# Description

This allows to use Uniswap V3 as oracle to calculate the GNP for Unlock.

Note that that part of the tests will now run on a mainnet fork only, as we dont want to have to redeploy Uniswap v3

```
export ALCHEMY_API_KEY=<xxx>
RUN_MAINNET_FORK=1 yarn hardhat test test/Unlock/uniswapValue.mainnet.js


Running a mainnet fork...

  Contract: Unlock / uniswapValue
    ✓ weth is set correctly already
    ✓ sets oracle address correctly
    A supported token (USDC)
      ✓ pricing is set correctly
      Purchase keys
        ✓ GDP went up by the expected ETH value
        ✓ a GDP tracking event has been emitted
    A unsupported token
      Purchase key
        ✓ GDP did not change
        ✓ a GDP tracking event has been emitted
    ETH
      Purchase key
        ✓ GDP went up by the keyPrice
        ✓ a GDP tracking event has been emitted

```

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by julien51** [comment]
> A supported token (USDC)

I think we already had USDC... though... I am not sure how to test that it is using _your_ new oracle and not the old one?

**Comment by clemsos** [comment]
oh ok then. I will use another one

**Comment by clemsos** [comment]
ok brought back the contracts here + using Shiba Inu instead of USDC

**Comment by julien51** [comment]
>   using Shiba Inu instead of USDC

😆

File: smart-contract-extensions/hardhat.config.js [hunk 1]
```diff
@@ -18,9 +23,31 @@ const settings = {
...
+
+// add mainnet fork -- if API key is present
+if (process.env.RUN_MAINNET_FORK) {
+  // eslint-disable-next-line no-console
+  console.log('Running a mainnet fork...')
+  const alchemyAPIKey = process.env.ALCHEMY_API_KEY
+  if (!alchemyAPIKey) {
+    throw new Error('Missing Alchemy API Key, couldnt run a mainnet fork')
+  }
+  const alchemyURL = `https://eth-mainnet.alchemyapi.io/v2/${alchemyAPIKey}`
```
Comment (julien51) [inline smart-contract-extensions/hardhat.config.js:36]: Let's just use our RPC https://rpc.unlock-protocol.com so we don't have to mess with API keys :)

File: smart-contracts/package.json [hunk 2]
```diff
@@ -21,8 +21,10 @@
     "@openzeppelin/hardhat-upgrades": "1.16.1",
     "@openzeppelin/test-helpers": "0.5.16",
     "@openzeppelin/upgrades-core": "1.12.0",
-    "@uniswap/v2-core": "1.0.1",
-    "@uniswap/v2-periphery": "1.1.0-beta.0",
+    "@uniswap/v2-core": "^1.0.1",
```
Comment (julien51) [inline smart-contracts/package.json:24]: Please fix dependencies...

Comment (julien51) [reply smart-contracts/package.json:24]: Do not use `^`

**Comment by clemsos** [comment]
I am at a pure loss of words here... the yarn install from uniswap repo branch still fails on CI with an unknown error 😓

```
#31 83.58 [output clipped, log limit 1MiB reached]
#31 ERROR: executor failed running [/bin/sh -c yarn install]: exit code: 1
------
 > [dev 13/21] RUN --mount=type=cache,target=/home/unlock/yarn-cache,uid=1000,gid=1000 yarn install:
------
executor failed running [/bin/sh -c yarn install]: exit code: 1
Service 'smart-contracts' failed to build : Build failed
```

File: smart-contracts/test/UniswapOracleV3/uniswapOracleV3.mainnet.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  [WETH, DAI],
+  [USDC, DAI],
+  // [ETH, USDC],
+].map(([one, two]) => [
+  // make sure we got correct checksum
+  ethers.utils.getAddress(one),
+  ethers.utils.getAddress(two),
+])
+
+// very unprecise way tp round up things...
```
Comment (julien51) [inline smart-contracts/test/UniswapOracleV3/uniswapOracleV3.mainnet.js:19]:
```suggestion
// very unprecise way to round up things...
```

**Review by julien51 — APPROVED** [review]

**Pull request merged**